### PR TITLE
Keep polishing the models

### PR DESCRIPTION
**Pull request by ramsayleung**
## Description

Keep polish the models by following the issues:

+ #127 
+ #149 

## Motivation and Context

To make `rspotify`'s API easier to use. 

+ [x] Types like FullAlbums, PageSimplifiedAlbums, etc exist as wrappers. This makes the API more complex. Instead, have that type exist solely in the API function that needs it, and then return a Vec<FullAlbum>. This drastically reduces the number of public API types. The wrapper objects need to be reduced: 
  - [x] `FullTracks`
  - [x] `CursorPageFullArtists`
  - [x] `SeversalSimplifiedShows`
  - [x] `DevicePayload`
  - [x] `AudioFeaturesPayload`
  - [x] `PageCategory`
  - [x] `PageSimpliedAlbums`
  - [x] `FullAlbums`
  - [x] `FullArtists`
+ [ ] `AudioAnalysisSection::mode` and `AudioFeatures::mode` are `f32s` but should be `Option<Mode>s` where `enum Mode { Major, Minor } `as it is more useful.
+ [ ] ~~`CursorBasedPage` should also not include the URLs.~~
+ [x] Rename `AudioFeatures.duration_ms` to duration, and change its type from `u32` to `std::time::Duration`.
+ [x] Rename `FullEpisode.duration_ms` to duration, and change its type from `u32` to `std::time::Duration`.
+ [x] Rename `SimplifiedEpisode.duration_ms` to duration, and change its type from `u32` to `std::time::Duration`.
+ [x] Rename `FullTrack.duration_ms` to duration, and change its type from `u32` to `std::time::Duration`.
+ [x] Rename `SimplifiedTrack.duration_ms` to duration, and change its type from `u32` to `std::time::Duration`.
+ [x] `SimplifiedEpisode::duration_ms` should be a `Duration`.
+ [x] `Offset::position` should be a `Duration`
+ [x] `CurrentlyPlayingContext::progress_ms` should be `progress`: `Option<Duration>`.
+ [x] `CurrentPlaybackContext::progress_ms` should be `progress`: `Option<Duration>`.
+ [x] `ResumePoint::resume_position_ms` should be a `Duration`.
+ [x] `CurrentlyPlayingContext::timestamp` should be a `DateTime<Utc>`
+ [x] `CurrentPlaybackContext::timestamp` should be a `DateTime<Utc>`
+ [x] ~~`SimplifiedPlayingContext::timestamp` should be a `DateTime<Utc>`~~(`SimplifiedPlayingContext` is useless, just remove it.)

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. 

Please also list any relevant details for your test configuration

- [x] Reduce wrapper object:
  - [x] `FullArtists`: `test_artist_related_artists`, `test_artists`
  - [x] `FullTracks`: `test_artist_top_tracks`, `test_tracks`
  - [x] `CursorPageFullArtists`: `test_current_user_followed_artists`
  - [x] `SeversalSimplifiedShows`: `test_get_seversal_shows`
  - [x] `DevicePayload`: `test_device`
  - [x] `AudioFeaturesPayload`: `test_audios_features`
  - [x] `PageCategory`: `test_categories`
  - [x] `PageSimpliedAlbums`: `test_new_releases`
  - [x] `FullAlbums`: `test_albums`
- [x] Rename `duration_ms` to `duration`, and change its type from `f32` to `std::time::Duration`
  - [x] `AudioFeatures`: `test_audio_features`
  - [x] `FullEpisode`: `test_full_episode`
  - [x] `SimplifiedEpisode`: `simplified_episode`
  - [x] `FullTrack`: `test_full_track` 
  - [x] `SimplifiedTrack`: `test_simplified_track` 
+ [x] `ResumePoint::resume_position_ms` should be a `Duration`: `test_resume_point`
- [x] `CurrentlyPlayingContext/CurrentlyPlaybackContext/SimplifiedPlayingContext::timestamp` should be a `DateTime<Utc>` 
  + [x] `CurrentlyPlayingContext`: `test_currently_playing_context`
  + [x] `CurrentPlaybackContext`: `test_current_playback_context`
+ [x] `Offset::position` should be a `Duration`: `test_offset`

PS: it seems the pull request template doesn't works, I just add the content manually, and I would like to open a new PR to test the pull request template individually

### CODE REVIEW

**Comment by marioortizmanero** [comment]
I came up with a different solution to remove repetitive work from the endpoints, only needing `.extract()`:

```rust
trait Extract {
    type InnerT;

    fn extract(self) -> Self::InnerT;
}

pub(in crate) struct BaseModel {
    onlyfield: Vec<String>
}

impl Extract for BaseModel {
    type InnerT = Vec<String>;

    fn extract(self) -> Self::InnerT {
        self.onlyfield
    }
}
```

It's not that much of a change, but at least I would make the simplified model public but only in crate. I don't like declaring it inside the endpoint. The `extract` part is just so that its usage is grouped up into a trait and so that it remains consistent, but it's not really necessary as it might overcomplicate things.

**Comment by marioortizmanero** [comment]
In fact, the `Extract` trait is trivial to implement for any struct, so it could have a macro:

```
trait Extract {
    type InnerT;

    fn extract(self) -> Self::InnerT;
}

macro_rules! simplified_model {
    ($vis:vis struct $name:ident { $attr_name:ident : $attr_type:ty }) => {
        #[derive(Debug)] // You can add any repetitive derives here
        $vis struct $name {
            $attr_name: $attr_type
        }

        impl Extract for $name {
            type InnerT = $attr_type;

            fn extract(self) -> Self::InnerT {
                self.$attr_name
            }
        }
    }
}

simplified_model! {
    pub(in crate) struct BaseModel {
        onlyfield: Vec<String>
    }
}
```

...which I've tested and works, but it's up to you if you consider it necessary as well.

**Comment by ramsayleung** [comment]
> I don't like declaring it inside the endpoint. The extract part is just so that its usage is grouped up into a consistent-to-use trait, but it's not really necessary as it might overcomplicate things.

Yes, I have to confess that declaring model inside the endpoints is not a good idea, since if multiple endpoints return a same wrapper object, we will define it repeatedly inside different endpoints. 

On the other side, I think define a new trait and macro to extra the valuable field from wrapper model is over-design, because wrapper objects mostly are a simple struct, it's easy to extract field with `Result.map` function.

So I personally prefer to redefine those wrapper objects with `(in crate)` in the `model` module, and extract the valuable field with `Result.map` function.

**Comment by ramsayleung** [comment]
@marioortizmanero I think this PR is ready to review, as for this TODO item: 

> `AudioAnalysisSection::mode` and `AudioFeatures::mode` are `f32s` but should be `Option<Mode>s` where `enum Mode { Major, Minor } `as it is more useful.

I have created an issue #163 to discuss that is there a better choice to implement it.

File: src/model/mod.rs [hunk 1]
```diff
@@ -17,7 +17,127 @@ pub mod show;
...
+    }
+    fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Duration::from_millis(v))
+    }
+}
+
+/// Deserialize `std::time::Duration` from millisecond(represented as u64)
```
Comment (marioortizmanero) [inline src/model/mod.rs:39]: These comments could have an extra space before the parenthesis. `millisecond` could be changed to `milliseconds` as well.

File: src/model/mod.rs [hunk 2]
```diff
@@ -17,7 +17,127 @@ pub mod show;
...
+
+/// Deserialize Unix millisecond timestamp to `DateTime<Utc>`
+pub(in crate) fn from_millisecond_timestamp<'de, D>(d: D) -> Result<DateTime<Utc>, D::Error>
+where
+    D: de::Deserializer<'de>,
+{
+    d.deserialize_u64(DateTimeVisitor)
+}
+
+/// Serialize DateTime<Utc>  to Unix millisecond timestamp
```
Comment (marioortizmanero) [inline src/model/mod.rs:89]: There's two spaces between `DateTime<Utc>` and `to`

**Review by marioortizmanero — APPROVED** [review]
LGTM! I haven't found any issues other than small grammatical nitpicks. Really great work :)

**Comment by ramsayleung** [comment]
> I haven't found any issues other than small grammatical nitpicks.

I would fix those problems later :)

**Pull request merged**